### DISCUSSION TITLE
list files/directories from datastore

### DISCUSSION THREAD

**Discussion by elungu3** [vmware/pyvmomi-community-samples#737] [Q&A]
Hello @liuqi @prziborowski 
Is there any method to list the files from a specified path in a specified datastore?
For example, I want to upload files to datastore, which I implemented and its working, but firstly I want to check if that file is already uploaded. Is there any method which I can use to list out all the files in a specific location and check if my file is there?
Thanks in advance!

**Comment by prziborowski** [comment]
Each datastore has a [browser](https://developer.vmware.com/apis/358/vim.host.DatastoreBrowser.html) that has [search](https://developer.vmware.com/apis/358/vim.host.DatastoreBrowser.html#search) and [searchSubFolders](https://developer.vmware.com/apis/358/vim.host.DatastoreBrowser.html#searchSubFolders) methods.
You have to pass in a [searchSpec](https://developer.vmware.com/apis/358/vim.host.DatastoreBrowser.SearchSpec.html) and getting that right can be tricky. The [query](https://developer.vmware.com/apis/358/vim.host.DatastoreBrowser.Query.html) property has many subclasses, I believe for returning certain properties to return, but if you want to look for just what files are there then [FolderQuery](https://developer.vmware.com/apis/358/vim.host.DatastoreBrowser.FolderQuery.html) is probably what you want. 
And if you are looking for a particular file, then I think `matchPattern` could just be the file you are interested in (e.g.  `[fileName]` as it is a list, so you don't want send in each character as an entry).

**Reply by elungu3** [reply]
Thank you for your quick reply.

I had a look over it, and I can't get this method working. Unfortunately I could not find any usage of that method or any example. Here is what I tried which failed:

    def filequery(self):
        for datastore in self.datastore:
            if datastore.summary.name == self.datastore_name:
                dc_brw = datastore.browser
                print(dc_brw) #this will print 'vim.host.DatastoreBrowser:datastoreBrowser-datastore-4139'
        folderq = dc_brw.FolderFileQuery('/testfolder') # this will fail AttributeError: 'vim.host.DatastoreBrowser' object has no attribute 'Query'

Could you please help me with this? It would be much appreciated. 
As an example, I have a folder in the root of a specific datastore called 'testfolder' and want to return a list with the names of all the files inside.

**Reply by prziborowski** [reply]
```
from pyVim.task import WaitForTask

def filequery(self):
    for datastore in self.datastore:
        if datastore.summary.name == self.datastore_name:
            dc_brw = datastore.browser
            print(dc_brw) #this will print 'vim.host.DatastoreBrowser:datastoreBrowser-datastore-4139'
            break
    searchSpec = vim.host.DatastoreBrowser.SearchSpec(
        query=vim.host.DatastoreBrowser.FolderQuery(),
        matchPattern=['*'], # Not positive if it needs to be set
    )
    search_task = dc_brw.Search(f'[{datastore.summary.name}] testfolder', searchSpec)
    WaitForTask(search_task)
    result = search_task.info.result
    print(result) # https://developer.vmware.com/apis/358/vim.host.DatastoreBrowser.SearchResults.html
```

**Reply by elungu3** [reply]
it looks like for this code provided I get the same error I was getting too: 
  File "/home/me/.local/lib/python3.7/site-packages/pyVmomi/VmomiSupport.py", line 1105, in CheckField
    for it in val:
TypeError: 'vim.host.DatastoreBrowser.FolderQuery' object is not iterable

Meanwhile, I used a GET method on the link I want to use PUT and then work out with the http type output, even tho this method is a headache, it works, but if I could get the FolderQuery method working it would be much easier and much appreciated. 
Thanks a lot!

**Reply by prziborowski** [reply]
Oh, my mistake, should be:
```
        query=[vim.host.DatastoreBrowser.FolderQuery()],
```
as the method takes an array of query and not just a single one.
